### PR TITLE
Screen editor

### LINKED ISSUE #21: [TASK] Create "Screen Editor" screen
Depends on #19 
Design:
Screenshot from 2020-03-21 10-48-44

Screenshot from 2020-03-21 10-50-47

### PR DESCRIPTION
**Pull request by ACmiester**
Har lavet en basic implementation af screen editoren.

Det der mangler er:
- Slet funktion
- Bedre UI i configuration delen
- Flere funktioner til configuration delen
- Flere widgets

Tænker at denne PR closes #21 da skærmen er lavet, men at vi så har et andet issue som er advance, med de forskellige ekstra ting vi skal have med.
Hvis ikke, så skriv, og så kan jeg kigge mere på det.

### CODE REVIEW

File: application/lib/model/Screen.dart [hunk 1]
```diff
@@ -6,7 +6,7 @@ class Screen implements Json {
 
   Screen.fromJson(Map<String, dynamic> json) {
     screenName = json['screenName'];
-    screenContent = jsonDecode(json['screenContent']);
+    screenContent = jsonDecode(json['screenContent'].toString().replaceAll('\'', '\"'));
```
Comment (sipman) [inline application/lib/model/Screen.dart]: why do you need to replace single quotes with double quotes ?

File: application/lib/providers/BaseApi.dart [hunk 2]
```diff
@@ -214,11 +214,12 @@ class BaseApi {
...
-    final String data = '''
+   final String data = '''
     {
-      "screenName": "$screenName",
-      "screenContent": "$screenContent"
-    }''';
+      \"screenName\": \"$screenName\",
+      \"screenContent\": \"$screenContent\"
+    }
+    ''';
```
Comment (sipman) [inline application/lib/providers/BaseApi.dart]: This looks a little bit iffy - why dont we extract this to a value class, then we have a reuseable object that is easy to change if the interface ever changes. This approach here, we have multiple issues down the line.

File: application/lib/view/screens/ScreenEditorScreen.dart [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+import 'package:font_awesome_flutter/font_awesome_flutter.dart';
+
+class ScreenEditorScreen extends BaseScreen {
+  ScreenEditorScreen(this.screen) {
+    for (var widget in screen.screenContent) {
+      screenContent.add(EditorScreenElement(
+          widgetType: widget['type'],
+          key: widget['key'],
+          position: int.parse(widget['position'])));
+    }
```
Comment (sipman) [inline application/lib/view/screens/ScreenEditorScreen.dart]: This will brake in the long run - how will you handle keys that one widget has that the other widgets dont have ?

TL;DR - Use factory pattern instead of this !

To fix this i would recommend that `EditorScreenElement` becomes an abstract class with a static build method and then you create a class per Widget type that inherits from `EditorScreenElement`. In this for-loop you can then send the entire json as `screenContent.add(EditorScreenElement.build(widget))` - The factory class will then check type to see which concrete class to new up, and on that you can have a `fromJson`-method that knows how to convert json into it own data properties..

File: application/lib/view/screens/ScreenEditorScreen.dart [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+  final ScreenBloc screenBloc = di.getDependency<ScreenBloc>();
+  final StreamController<EditorScreenElement> selectedElement =
+      StreamController<EditorScreenElement>.broadcast();
+
+  final Screen screen;
+
+  final List<EditorScreenElement> screenContent = [];
+
+  final List<String> widgets = ['Text', 'Flat button', 'Label'];
```
Comment (sipman) [inline application/lib/view/screens/ScreenEditorScreen.dart]: This should get injected and not coded in here, we need some kind of pattern that can decouple this.

File: application/lib/view/screens/ScreenEditorScreen.dart [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+              content.add(
+                const SizedBox(
+                  height: 16,
+                ),
+              );
+            }
+            return Padding(
+              padding: const EdgeInsets.only(top: 8.0),
+              child: Column(children: content),
+            );
```
Comment (sipman) [inline application/lib/view/screens/ScreenEditorScreen.dart]: I believe you can use higher-order functions to clean this up a bit.

File: application/lib/view/screens/ScreenSelectionScreen.dart [hunk 6]
```diff
@@ -99,36 +99,44 @@ class ScreenSelectionScreen extends BaseScreen {
...
-        return FlatButton(onPressed: null, child: Text(widget['value']));
+        return const Text('Text');
+      case 'label':
+        return const Text('Label');
+      case 'flat button':
+        return const FlatButton(onPressed: null, child: Text('Flat button'));
       default:
-        throw Exception('Widget not found. Either add it or ensure no typos');
+        throw Exception(widget['type'] +' not found. Either add it or ensure no typos');
     }
```
Comment (sipman) [inline application/lib/view/screens/ScreenSelectionScreen.dart]: Duplicated string (widget types) - I would recommend that you instead of string use a enum - such that misspelling comes a non-issue. And the create a different class or pattern that have all the information about widgets and how to create them - this can then be injected into who ever needs it!

**Review by sipman — CHANGES_REQUESTED** [review]

**Review by sipman — DISMISSED** [review]

**Review by sipman — APPROVED** [review]
